projects/hipdnn/plugins/miopen_legacy_plugin/engines/plans/MiopenBatchnormPlanBuilder.cpp

```diff
@@ -8,33 +8,235 @@
+    }
+
+    // Verify that different BN operations use shared inputs where applicable
+    if(bnBwdAttr.x_tensor_uid() != bnInfAttr.x_tensor_uid())
+    {
+        throw hipdnn_plugin::HipdnnPluginException(
+            HIPDNN_PLUGIN_STATUS_BAD_PARAM,
+            "Batchnorm backward must use the same X tensor as batchnorm inference");
+    }
+
+    if(bnBwdAttr.mean_tensor_uid().has_value()
+       && bnBwdAttr.mean_tensor_uid().value() != bnInfAttr.mean_tensor_uid())
+    {
+        throw hipdnn_plugin::HipdnnPluginException(
+            HIPDNN_PLUGIN_STATUS_BAD_PARAM,
+            "Batchnorm backward must use the same mean tensor as batchnorm inference");
+    }
+
+    if(bnBwdAttr.inv_variance_tensor_uid().has_value()
+       && bnBwdAttr.inv_variance_tensor_uid().value() != bnInfAttr.inv_variance_tensor_uid())
+    {
+        throw hipdnn_plugin::HipdnnPluginException(
+            HIPDNN_PLUGIN_STATUS_BAD_PARAM,
+            "Batchnorm backward must use the same inv_variance tensor as batchnorm inference");
+    }
+
+    if(bnBwdAttr.scale_tensor_uid() != bnInfAttr.scale_tensor_uid())
+    {
+        throw hipdnn_plugin::HipdnnPluginException(
+            HIPDNN_PLUGIN_STATUS_BAD_PARAM,
+            "Batchnorm backward must use the same scale tensor as batchnorm inference");
+    }
+
+    // Check for virtual tensors
+    const auto& bnInfTensorX
+        = miopen_utils::findTensorAttributes(tensorMap, bnInfAttr.x_tensor_uid());
+    const auto& bnInfTensorMean
+        = miopen_utils::findTensorAttributes(tensorMap, bnInfAttr.mean_tensor_uid());
+    const auto& bnInfTensorInvVar
+        = miopen_utils::findTensorAttributes(tensorMap, bnInfAttr.inv_variance_tensor_uid());
+    const auto& bnInfTensorScale
+        = miopen_utils::findTensorAttributes(tensorMap, bnInfAttr.scale_tensor_uid());
+    const auto& bnInfTensorBias
+        = miopen_utils::findTensorAttributes(tensorMap, bnInfAttr.bias_tensor_uid());
+    const auto& bnInfTensorY
+        = miopen_utils::findTensorAttributes(tensorMap, bnInfAttr.y_tensor_uid());
+
+    if(bnInfTensorX.virtual_() || bnInfTensorMean.virtual_() || bnInfTensorInvVar.virtual_()
+       || bnInfTensorScale.virtual_() || bnInfTensorBias.virtual_() || !bnInfTensorY.virtual_())
+    {
+        throw hipdnn_plugin::HipdnnPluginException(
+            HIPDNN_PLUGIN_STATUS_BAD_PARAM,
+            "Batchnorm inference input tensors must be non-virtual, output tensor must be virtual");
+    }
+
+    const auto& actTensorIn0
+        = miopen_utils::findTensorAttributes(tensorMap, actAttr.in_0_tensor_uid());
+    const auto& actTensorOut
+        = miopen_utils::findTensorAttributes(tensorMap, actAttr.out_0_tensor_uid());
+
+    const auto& actBnInputTensor
+        = (actAttr.in_0_tensor_uid() == bnInfAttr.y_tensor_uid())
+              ? actTensorIn0
+              : miopen_utils::findTensorAttributes(tensorMap, actAttr.in_1_tensor_uid().value());
```

Comment:
Nit: I'd change this to 
```suggestion
    const auto& actTensorOut
        = miopen_utils::findTensorAttributes(tensorMap, actAttr.out_0_tensor_uid());

    auto actTensorInUid = (actAttr.in_0_tensor_uid() == bnInfAttr.y_tensor_uid()) 
        ? actAttr.in_0_tensor_uid(),value()
        : actAttr.in_1_tensor_uid().value();
        
    const auto& actBnInputTensor = miopen_utils::findTensorAttributes(tensorMap, actTensorInUid);
```
to avoid having the intermediate tensor you don't use elsewhere